File: 70-tools/gftd/gftd/nono.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/etzhayyim/root/70-tools/gftd/gftd/db"
)

// nonoManifest represents a nono-manifest.jsonld capability provider definition.
type nonoManifest struct {
	Context          string            `json:"@context"`
	ID               string            `json:"@id"`
	Name             string            `json:"name"`
	Nanoid           string            `json:"nanoid"`
	Type             string            `json:"type"` // always "nono"
	Bindings         []string          `json:"bindings"`
	PrimitiveBackend []string          `json:"primitiveBackend,omitempty"`
	Skills           []nonoSkill       `json:"skills,omitempty"`
	Governance       *governanceConfig `json:"governance,omitempty"`
}

type nonoSkill struct {
	NSID         string `json:"nsid"`
	Description  string `json:"description"`
	InputSchema  any    `json:"inputSchema,omitempty"`
	OutputSchema any    `json:"outputSchema,omitempty"`
}

// runNono implements `gftd nono` — capability Worker lifecycle.
//
// Usage:
//
//	gftd nono build  [-dir <path>]   # esbuild + validation
//	gftd nono deploy [-dir <path>]   # CF Worker deploy + skill registration
//	gftd nono list                   # list all nono capability Workers
//	gftd nono skills <nanoid>        # list skills provided by a nono
func runNono(args []string) error {
	if len(args) == 0 {
		fmt.Println(`gftd nono — capability Worker lifecycle

SUBCOMMANDS:
  build   [-dir <path>]    Build nono Worker (esbuild + validation)
  deploy  [-dir <path>]    Deploy nono Worker + register skills
  list                     List all nono capability Workers
  skills  <nanoid|name>    List skills provided by a nono`)
		return nil
	}

	switch args[0] {
	case "build":
		return runNonoBuild(args[1:])
	case "deploy":
		return runNonoDeploy(args[1:])
	case "list":
		return runNonoList(args[1:])
	case "skills":
		return runNonoSkills(args[1:])
	default:
		return fmt.Errorf("unknown nono subcommand: %s. Available: build, deploy, list, skills", args[0])
	}
}

// runNonoBuild builds a nono Worker (validates + esbuild).
func runNonoBuild(args []string) error {
	fs := flag.NewFlagSet("nono build", flag.ContinueOnError)
	dir := fs.String("dir", ".", "component directory")
	if err := fs.Parse(args); err != nil {
		return err
	}

	compDir, err := filepath.Abs(*dir)
	if err != nil {
		return err
	}

	// Must have magatama.jsonld (nono Workers still use magatama.jsonld for deploy config)
	cfg, err := readMagatamaJSONLD(compDir)
	if err != nil {
		return fmt.Errorf("magatama.jsonld required: %w", err)
	}

	fmt.Printf("[nono build] %s (nanoid: %s)\n", cfg.Name, cfg.Nanoid)

	// Validate src/app.ts exists
	appTs := filepath.Join(compDir, "src", "app.ts")
	if _, err := os.Stat(appTs); err != nil {
		return fmt.Errorf("src/app.ts not found — nono Workers require app.ts entry point")
	}
	fmt.Printf("  ✓ src/app.ts exists\n")

	// Check for CF bindings that justify nono status
	nonoManifestPath := filepath.Join(compDir, "nono-manifest.jsonld")
	if _, err := os.Stat(nonoManifestPath); err == nil {
		data, _ := os.ReadFile(nonoManifestPath)
		var nm nonoManifest
		if json.Unmarshal(data, &nm) == nil && len(nm.Bindings) > 0 {
			fmt.Printf("  ✓ nono-manifest.jsonld (bindings: %s)\n", strings.Join(nm.Bindings, ", "))
		}
	} else {
		fmt.Printf("  ℹ no nono-manifest.jsonld (optional — bindings documented in wrangler.jsonc)\n")
	}

	// Delegate to standard build
	fmt.Printf("\n[nono build] Running standard build...\n")
	return runBuild([]string{"-dir", compDir})
}

// runNonoDeploy deploys a nono Worker and registers skills.
func runNonoDeploy(args []string) error {
	fs := flag.NewFlagSet("nono deploy", flag.ContinueOnError)
	dir := fs.String("dir", ".", "component directory")
	skipSkills := fs.Bool("skip-skills", false, "skip skill registration after deploy")
	if err := fs.Parse(args); err != nil {
		return err
	}

	compDir, err := filepath.Abs(*dir)
	if err != nil {
		return err
	}

	cfg, err := readMagatamaJSONLD(compDir)
	if err != nil {
		return fmt.Errorf("magatama.jsonld required: %w", err)
	}

	appID := cfg.AppID()
	fmt.Printf("[nono deploy] Phase 1: Worker deploy\n")
	fmt.Printf("  name: %s, nanoid: %s\n", cfg.Name, appID)

	// Standard Worker deploy (wrangler)
	if err := deployWorker(cfg, compDir, deployWorkerOptions{}); err != nil {
		return fmt.Errorf("worker deploy failed: %w", err)
	}

	if *skipSkills {
		fmt.Printf("\n[nono deploy] Skill registration skipped (--skip-skills)\n")
		return nil
	}

	// Phase 2: Register nono skills to graph
	fmt.Printf("\n[nono deploy] Phase 2: Skill registration\n")

	nonoManifestPath := filepath.Join(compDir, "nono-manifest.jsonld")
	if _, err := os.Stat(nonoManifestPath); err != nil {
		fmt.Printf("  ℹ no nono-manifest.jsonld — skill registration skipped\n")
		fmt.Printf("  ℹ skills are auto-discovered from sdk.app.command() via MCP tools/list\n")
		return nil
	}

	data, err := os.ReadFile(nonoManifestPath)
	if err != nil {
		return fmt.Errorf("read nono-manifest.jsonld: %w", err)
	}

	var nm nonoManifest
	if err := json.Unmarshal(data, &nm); err != nil {
		return fmt.Errorf("parse nono-manifest.jsonld: %w", err)
	}

	token := resolveGFTDToken()
	if token == "" {
		fmt.Printf("  ⚠ no auth token — skill registration requires authentication\n")
		return nil
	}

	// Register nono manifest to graph
	pdsBase := resolvePDSBaseURL()
	regBody, _ := json.Marshal(map[string]any{
		"@context":         "https://etzhayyim.com/ns/nono/v1",
		"@id":              nm.ID,
		"name":             nm.Name,
		"nanoid":           nm.Nanoid,
		"type":             "nono",
		"bindings":         nm.Bindings,
		"primitiveBackend": nm.PrimitiveBackend,
		"skills":           nm.Skills,
		"status":           "active",
	})

	req, err := http.NewRequest("POST", pdsBase+"/xrpc/ai.gftd.actor.registerManifest", strings.NewReader(string(regBody)))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	setAuthHeaders(req)

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("  ⚠ skill registration failed: %v\n", err)
		return nil
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		fmt.Printf("  ⚠ skill registration HTTP %d\n", resp.StatusCode)
		return nil
	}

	fmt.Printf("  ✓ nono registered: %s (%d bindings, %d skills)\n",
		nm.Name, len(nm.Bindings), len(nm.Skills))

	if len(nm.PrimitiveBackend) > 0 {
		fmt.Printf("  ✓ primitive backend: %s\n", strings.Join(nm.PrimitiveBackend, ", "))
	}

	for _, skill := range nm.Skills {
		fmt.Printf("    - %s: %s\n", skill.NSID, skill.Description)
	}

	fmt.Printf("\n[nono deploy] Complete: %s deployed as capability Worker\n", nm.Name)
	return nil
}

// runNonoList lists all registered nono capability Workers.
func runNonoList(args []string) error {
	fs := flag.NewFlagSet("nono list", flag.ContinueOnError)
	jsonOut := fs.Bool("json", false, "JSON output")
	if err := fs.Parse(args); err != nil {
		return err
	}

	token := resolveGFTDToken()
	if token == "" {
		// Offline mode: scan 60-apps for nono candidates
		return nonoListOffline()
	}

	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
	defer cancel()
	result, err := db.RawQuery(ctx, `
		SELECT did, name, nanoid, bindings
		FROM vertex_actor_manifest
		WHERE type = 'nono' AND status = 'active'
		ORDER BY name
		LIMIT 50
	`)
	if err != nil {
		fmt.Fprintf(os.Stderr, "  database query failed, falling back to offline scan\n")
		return nonoListOffline()
	}

	if *jsonOut {
		enc := json.NewEncoder(os.Stdout)
		enc.SetIndent("", "  ")
		return enc.Encode(result)
	}

	rows := result.Rows
	if len(rows) == 0 {
		fmt.Println("No nono Workers registered in database. Showing offline scan:")
		return nonoListOffline()
	}

	fmt.Printf("%-20s %-12s %-30s %s\n", "NAME", "NANOID", "BINDINGS", "DID")
	fmt.Printf("%-20s %-12s %-30s %s\n", "----", "------", "--------", "---")
	for _, r := range rows {
		fmt.Printf("%-20s %-12s %-30s %s\n",
			r["name"], r["nanoid"], r["bindings"], r["did"])
	}
	return nil
}

// nonoListOffline lists known nono Workers from hardcoded registry.
func nonoListOffline() error {
	nonos := []struct {
		name    string
		binding string
		backend string
		nanoid  string
	}{
		{"llm", "env.AI (Workers AI)", "agent.chat", "llm8cf4ai"},
		{"site", "HEADLESS_BROWSER", "browser.fetch", "w3bpg001"},
		{"livecam", "Murakumo CoreML fleet", "-", "zio3cb3r"},
		{"mangaka", "env.AI (flux-1-schnell)", "-", "mng4k4x1"},
		{"briefing", "WebRTC (kami-rtc)", "-", "w3olw1pf"},
		{"stripe", "Stripe API + Webhooks", "-", "st4rp301"},
		{"watashi", "Native CGEvent/Win32", "-", "mfbtsuyc"},
		{"sense", "Camera/LiDAR/WiFi/BT/Mic", "-", "sense"},
		{"pptx", "wgpu WebGPU rendering", "-", "t53br1o0"},
		{"xlsx", "HTML DOM grid", "-", "il0ndq6a"},
	}

	fmt.Printf("%-15s %-12s %-30s %s\n", "NAME", "NANOID", "CF BINDING", "PRIMITIVE BACKEND")
	fmt.Printf("%-15s %-12s %-30s %s\n", "----", "------", "----------", "----------------")
	for _, n := range nonos {
		fmt.Printf("%-15s %-12s %-30s %s\n", n.name, n.nanoid, n.binding, n.backend)
	}
	fmt.Printf("\nTotal: %d nono capability Workers\n", len(nonos))
	return nil
}

// runNonoSkills lists skills provided by a specific nono.
func runNonoSkills(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: gftd nono skills <nanoid|name>")
	}
	target := args[0]

	token := resolveGFTDToken()
	if token == "" {
		return fmt.Errorf("no auth token — run: gftd auth login")
	}

	// Resolve DID from nanoid or name
	did := target
	if !strings.HasPrefix(did, "did:") {
		did = "did:web:" + target + ".etzhayyim.com"
	}

	pdsBase := resolvePDSBaseURL()

	// Try to fetch nono manifest from graph
	body := fmt.Sprintf(`{"did":"%s"}`, did)
	req, err := http.NewRequest("POST", pdsBase+"/xrpc/ai.gftd.actor.getManifest", strings.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	setAuthHeaders(req)

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("query failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == 404 {
		// Fallback: query MCP tools/list from the nono Worker directly
		fmt.Printf("No manifest in graph. Querying MCP tools/list from %s...\n\n", target)
		return nonoSkillsFromMCP(target)
	}

	var manifest map[string]any
	json.NewDecoder(resp.Body).Decode(&manifest)

	skills, _ := manifest["skills"].([]any)
	primitiveBackend, _ := manifest["primitiveBackend"].([]any)
	bindings, _ := manifest["bindings"].([]any)

	name, _ := manifest["name"].(string)
	fmt.Printf("Nono: %s\n", name)
	if len(bindings) > 0 {
		fmt.Printf("Bindings: %v\n", bindings)
	}
	if len(primitiveBackend) > 0 {
		fmt.Printf("Primitive Backend: %v\n", primitiveBackend)
	}
	fmt.Printf("\nSkills (%d):\n", len(skills))
	for _, s := range skills {
		skill, _ := s.(map[string]any)
		fmt.Printf("  %-50s %s\n", skill["nsid"], skill["description"])
	}
	return nil
}

// nonoSkillsFromMCP queries MCP tools/list from a nono Worker endpoint.
func nonoSkillsFromMCP(nanoid string) error {
	mcpURL := "https://mcp.etzhayyim.com/mcp"
	body := fmt.Sprintf(`{"jsonrpc":"2.0","id":1,"method":"tools/list","params":{"nanoid":"%s"}}`, nanoid)
	req, err := http.NewRequest("POST", mcpURL, strings.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("MCP query failed: %w", err)
	}
	defer resp.Body.Close()

	var result map[string]any
	json.NewDecoder(resp.Body).Decode(&result)

	mcpResult, _ := result["result"].(map[string]any)
	tools, _ := mcpResult["tools"].([]any)

	fmt.Printf("MCP Tools (%d):\n", len(tools))
	for _, t := range tools {
		tool, _ := t.(map[string]any)
		fmt.Printf("  %-50s %s\n", tool["name"], tool["description"])
	}
	return nil
}

File: 70-tools/etzhayyim-py/src/etzhayyim/nono.py
"""nono — Nono capability worker lifecycle management.

Nono workers provide NSID-bound capability primitives.
Reads nono-manifest.jsonld files from workspace.
"""

from __future__ import annotations

import json
import sys
from dataclasses import dataclass, field
from pathlib import Path

import click
import httpx

from .authn import _load_auth
from .projector import resolve_pds
from .shannon import _resolve_root


@dataclass
class NonoManifest:
    nanoid: str
    name: str
    bindings: list[str] = field(default_factory=list)
    skills: list[dict] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {"nanoid": self.nanoid, "name": self.name,
                "bindings": self.bindings, "skills": self.skills}


def _load_manifests(ws: Path) -> list[NonoManifest]:
    manifests = []
    for p in ws.rglob("nono-manifest.jsonld"):
        try:
            data = json.loads(p.read_text(errors="replace"))
        except (OSError, json.JSONDecodeError):
            continue
        nanoid = data.get("nanoid", "")
        if nanoid:
            manifests.append(NonoManifest(
                nanoid=nanoid,
                name=data.get("name", ""),
                bindings=data.get("bindings", []),
                skills=data.get("skills", []),
            ))
    return manifests


def _auth_headers() -> dict:
    auth = _load_auth()
    tok = auth.get("accessJwt") or auth.get("access_token") or ""
    if not tok:
        click.echo("not signed in — run: etzhayyim authn signin", err=True)
        sys.exit(1)
    return {"Authorization": f"Bearer {tok}", "Content-Type": "application/json"}


@click.group("nono", invoke_without_command=True)
@click.option("--workspace-dir", default=None)
@click.option("--json", "json_out", is_flag=True, default=False)
@click.pass_context
def nono(ctx: click.Context, workspace_dir: str | None, json_out: bool) -> None:
    """Nono capability worker lifecycle management."""
    if ctx.invoked_subcommand is not None:
        return
    ws = _resolve_root(workspace_dir)
    manifests = _load_manifests(ws)
    if json_out:
        click.echo(json.dumps([m.to_dict() for m in manifests], ensure_ascii=False, indent=2))
    else:
        click.echo(f"nono workers: {len(manifests)}")
        for m in manifests:
            click.echo(f"  {m.nanoid}  {m.name}  skills={len(m.skills)}")


@nono.command("list")
@click.option("--workspace-dir", default=None)
@click.option("--pds", default=None)
@click.option("--json", "json_out", is_flag=True, default=False)
def nono_list(workspace_dir: str | None, pds: str | None, json_out: bool) -> None:
    """List nono workers (local manifests + remote registry)."""
    ws = _resolve_root(workspace_dir)
    local = _load_manifests(ws)
    if json_out:
        click.echo(json.dumps([m.to_dict() for m in local], ensure_ascii=False, indent=2))
    else:
        for m in local:
            click.echo(f"  {m.nanoid}  {m.name}  bindings={','.join(m.bindings[:3])}")


@nono.command("inspect")
@click.argument("nanoid")
@click.option("--workspace-dir", default=None)
@click.option("--json", "json_out", is_flag=True, default=False)
def nono_inspect(nanoid: str, workspace_dir: str | None, json_out: bool) -> None:
    """Inspect a nono worker manifest."""
    ws = _resolve_root(workspace_dir)
    manifests = _load_manifests(ws)
    found = next((m for m in manifests if m.nanoid == nanoid), None)
    if not found:
        raise click.ClickException(f"nono manifest not found: {nanoid}")
    if json_out:
        click.echo(json.dumps(found.to_dict(), ensure_ascii=False, indent=2))
    else:
        click.echo(f"nanoid: {found.nanoid}")
        click.echo(f"name: {found.name}")
        click.echo(f"bindings: {found.bindings}")
        for s in found.skills:
            click.echo(f"  skill: {s.get('nsid', '')}  {s.get('description', '')[:60]}")


@nono.command("build")
@click.argument("nanoid", required=False, default=None)
@click.option("--dir", "comp_dir", default=None,
              help="Component directory (default: search workspace for nanoid)")
@click.option("--workspace-dir", default=None)
@click.option("--dry-run", is_flag=True, default=False)
def nono_build(nanoid: str | None, comp_dir: str | None,
               workspace_dir: str | None, dry_run: bool) -> None:
    """Build a nono worker (esbuild + validation)."""
    import subprocess

    # Resolve component directory
    if comp_dir:
        target_dir = Path(comp_dir).resolve()
    elif nanoid:
        ws = _resolve_root(workspace_dir)
        target_dir = None
        for manifest_path in ws.rglob("nono-manifest.jsonld"):
            try:
                data = json.loads(manifest_path.read_text(errors="replace"))
            except (OSError, json.JSONDecodeError):
                continue
            if data.get("nanoid") == nanoid:
                target_dir = manifest_path.parent
                break
        if target_dir is None:
            if dry_run:
                click.echo(f"[nono build] dry-run: {nanoid}  (nono not found in workspace)")
                return
            raise click.ClickException(
                f"nono not found: {nanoid}. Use --dir to specify the component directory.")
    else:
        target_dir = Path.cwd()

    wrangler_jsonc = target_dir / "wrangler.jsonc"
    pkg = target_dir / "package.json"

    if dry_run:
        click.echo(f"[nono build] dry-run: {nanoid or target_dir.name}")
        click.echo(f"  dir: {target_dir}")
        click.echo(f"  wrangler.jsonc: {'found' if wrangler_jsonc.exists() else 'missing'}")
        click.echo(f"  package.json: {'found' if pkg.exists() else 'missing'}")
        return

    click.echo(f"[nono build] Building {nanoid or target_dir.name}...")
    # Use npm/pnpm build if package.json exists, otherwise fall back to wrangler
    build_cmd = None
    if pkg.exists():
        pkg_data = json.loads(pkg.read_text(errors="replace"))
        if "build" in pkg_data.get("scripts", {}):
            build_cmd = ["pnpm", "run", "build"] if (target_dir / "pnpm-lock.yaml").exists() \
                        else ["npm", "run", "build"]
    if build_cmd is None and wrangler_jsonc.exists():
        build_cmd = ["npx", "wrangler", "deploy", "--dry-run", "--outdir", "dist"]
    if build_cmd is None:
        raise click.ClickException(
            f"no package.json scripts.build or wrangler.jsonc found in {target_dir}")

    r = subprocess.run(build_cmd, cwd=str(target_dir))
    if r.returncode != 0:
        raise click.ClickException(f"build failed (exit {r.returncode})")
    click.echo(f"[nono build] Complete: {target_dir.name}")


@nono.command("deploy")
@click.argument("nanoid")
@click.option("--dir", "comp_dir", default=None,
              help="Component directory (default: search workspace for nanoid)")
@click.option("--pds", default=None)
@click.option("--skip-skills", is_flag=True, default=False,
              help="Skip XRPC skill registration after deploy")
@click.option("--dry-run", is_flag=True, default=False)
@click.option("--workspace-dir", default=None)
def nono_deploy(nanoid: str, comp_dir: str | None, pds: str | None,
                skip_skills: bool, dry_run: bool, workspace_dir: str | None) -> None:
    """Deploy a nono worker (wrangler deploy + XRPC skill registration)."""
    import subprocess

    # Resolve component directory
    if comp_dir:
        target_dir = Path(comp_dir).resolve()
    else:
        # Search workspace for wrangler.jsonc near a nono-manifest with matching nanoid
        ws = _resolve_root(workspace_dir)
        target_dir = None
        for manifest_path in ws.rglob("nono-manifest.jsonld"):
            try:
                data = json.loads(manifest_path.read_text(errors="replace"))
            except (OSError, json.JSONDecodeError):
                continue
            if data.get("nanoid") == nanoid:
                target_dir = manifest_path.parent
                break
        if target_dir is None:
            # Fall back to searching by wrangler.jsonc dir that contains nanoid in name
            for wp in ws.rglob("wrangler.jsonc"):
                if nanoid in wp.parent.name:
                    target_dir = wp.parent
                    break
        if target_dir is None:
            if dry_run:
                click.echo(f"[nono deploy] dry-run: {nanoid}  (nono not found in workspace)")
                return
            raise click.ClickException(
                f"nono not found: {nanoid}. Use --dir to specify the component directory."
            )

    wrangler_jsonc = target_dir / "wrangler.jsonc"
    manifest_path = target_dir / "nono-manifest.jsonld"

    if dry_run:
        click.echo(f"[nono deploy] dry-run: {nanoid}")
        click.echo(f"  dir: {target_dir}")
        click.echo(f"  wrangler.jsonc: {'found' if wrangler_jsonc.exists() else 'missing'}")
        click.echo(f"  nono-manifest.jsonld: {'found' if manifest_path.exists() else 'missing'}")
        return

    # Phase 1: wrangler deploy
    click.echo(f"[nono deploy] Phase 1: Worker deploy  nanoid={nanoid}")
    if not wrangler_jsonc.exists():
        raise click.ClickException(f"wrangler.jsonc not found in {target_dir}")

    r = subprocess.run(["npx", "wrangler", "deploy"], cwd=str(target_dir))
    if r.returncode != 0:
        raise click.ClickException(f"wrangler deploy failed (exit {r.returncode})")

    if skip_skills:
        click.echo("\n[nono deploy] Skill registration skipped (--skip-skills)")
        return

    # Phase 2: XRPC skill registration
    click.echo("\n[nono deploy] Phase 2: Skill registration")
    if not manifest_path.exists():
        click.echo("  no nono-manifest.jsonld — skill registration skipped")
        click.echo("  skills are auto-discovered from sdk.app.command() via MCP tools/list")
        return

    try:
        nm_data = json.loads(manifest_path.read_text())
    except (OSError, json.JSONDecodeError) as e:
        click.echo(f"  parse nono-manifest.jsonld failed: {e}", err=True)
        return

    auth = _load_auth()
    token = auth.get("accessJwt") or auth.get("access_token") or ""
    if not token:
        click.echo("  no auth token — skill registration requires authentication")
        return

    pds_url = (pds or resolve_pds()).rstrip("/")
    reg_body = {
        "@context": "https://etzhayyim.com/ns/nono/v1",
        "@id": nm_data.get("@id", ""),
        "name": nm_data.get("name", ""),
        "nanoid": nm_data.get("nanoid", nanoid),
        "type": "nono",
        "bindings": nm_data.get("bindings", []),
        "primitiveBackend": nm_data.get("primitiveBackend", []),
        "skills": nm_data.get("skills", []),
        "status": "active",
    }
    try:
        resp = httpx.post(
            f"{pds_url}/xrpc/com.etzhayyim.actor.registerManifest",
            json=reg_body,
            headers={"Authorization": f"Bearer {token}", "Content-Type": "application/json"},
            timeout=15,
        )
        if resp.status_code >= 400:
            click.echo(f"  skill registration HTTP {resp.status_code}")
        else:
            bindings = nm_data.get("bindings", [])
            skills = nm_data.get("skills", [])
            click.echo(f"  ✓ nono registered: {nm_data.get('name', nanoid)} "
                       f"({len(bindings)} bindings, {len(skills)} skills)")
    except httpx.HTTPError as e:
        click.echo(f"  skill registration failed: {e}", err=True)

    click.echo(f"\n[nono deploy] Complete: {nm_data.get('name', nanoid)} deployed as capability Worker")


@nono.command("skills")
@click.argument("nanoid")
@click.option("--workspace-dir", default=None)
@click.option("--json", "json_out", is_flag=True, default=False)
def nono_skills(nanoid: str, workspace_dir: str | None, json_out: bool) -> None:
    """List skills of a nono worker."""
    ws = _resolve_root(workspace_dir)
    manifests = _load_manifests(ws)
    found = next((m for m in manifests if m.nanoid == nanoid), None)
    if not found:
        raise click.ClickException(f"nono manifest not found: {nanoid}")
    if json_out:
        click.echo(json.dumps(found.skills, ensure_ascii=False, indent=2))
    else:
        for s in found.skills:
            click.echo(f"  {s.get('nsid', '')}  {s.get('description', '')[:60]}")
